packages/pds/src/image/server.ts

```diff
@@ -0,0 +1,105 @@
+import fs from 'fs/promises'
+import path from 'path'
+import express, { ErrorRequestHandler, NextFunction } from 'express'
+import { InvalidRequestError, XRPCError } from '@atproto/xrpc-server'
+import { BadPathError, ImageUriBuilder } from './uri'
+import log from './logger'
+import { Readable } from 'stream'
+import { SharpImageProcessor } from './sharp'
+import { forwardStreamErrors, formatsToMimes, Options } from './util'
+
+export class ImageProcessingServer {
+  app = express()
+  processor = new SharpImageProcessor()
+  uriBuilder: ImageUriBuilder
+
+  constructor(
+    protected salt: Uint8Array,
+    protected key: Uint8Array,
+    protected storage: BlobStorage,
+  ) {
+    this.uriBuilder = new ImageUriBuilder(salt, key)
+    this.app.get('*', this.handler.bind(this))
+    this.app.use(errorMiddleware)
+  }
+
+  async handler(
+    req: express.Request,
+    res: express.Response,
+    next: NextFunction,
+  ) {
+    try {
+      const path = req.path
+      const options = this.uriBuilder.getVerifiedOptions(path)
+      const imageStream = await this.storage.get(options.fileId)
+      const processedImage = await this.processor.resize(imageStream, options)
+      res.statusCode = 200
+      res.setHeader('content-type', getMime(options.format))
+      res.setHeader('cache-control', `public, max-age=31536000`) // 1 year
+      forwardStreamErrors(processedImage, res)
+      return (
+        processedImage
+          // @NOTE sharp does emit this in time to be set as a header
+          .once('info', (info) => res.setHeader('content-length', info.size))
+          .pipe(res)
+      )
+    } catch (err: unknown) {
+      if (err instanceof BadPathError) {
+        return next(new InvalidRequestError(err.message))
+      }
+      if (err instanceof BlobNotFoundError) {
+        return next(new XRPCError(404, 'Image not found', 'NotFound'))
```

Comment:
Can't use 404 for XRPC error code 🙃

Reply:
Was thinking it could be acceptable here since this is a non-xrpc service, and I believe it matches imgproxy's behavior.  One awkward part is that I'm using XRPC errors, but it's only because they make for a nice http error interface—maybe I oughtta switch them to [http-errors](https://github.com/jshttp/http-errors) to avoid the mixed signals.

Reply:
Huh this actually brings up an interesting problem for XRPC -- if it doesn't support 404s as standard behaviors, it might have problems with proxies that are expecting 404s to indicate not found.

I wonder if the solution here is to use 404s in XRPC but differentiate between "lack of support" and "resource not found" by the response body? If the response body is a valid XRPC error object, we know it's the latter. Not sure why that didn't occur to me before!

Reply:
I have a feeling we'll eventually need to grapple with it one way or another, and I like that idea 👍

Reply:
yeah @pfrazee we actually briefly riffed on that yesterday^^
404 in HTTP already kinda means two things: "route not implemented" & "resource not found" & we're actually at an advantage with XRPC because we do have structured responses/errors